### DISCUSSION TITLE
Struggling with ssl on a debian docker container (python:3.10-slim-buster)

### DISCUSSION THREAD

**Discussion by Sammyalhashe** [planetscale/discussion#317] [Q&A] [answered]
Essentially I am attempting to connect to planetscale using a python client running on a debian container. I was originally looking at the following blog post before I got derailed: https://planetscale.com/blog/connect-to-a-mysql-database-in-python.

Currrently, I am struggling when attempting to connect:
```MySQL client library does not support ssl_mode specification```

After looking over the internet it appears that my issue is that mysql client that I have installed does not support ssl? That being said I have no idea how I can do this or get around this right now. Any tips/tricks to do so would be great.

In my container, I an running:
```docker
RUN apt-get -y update \
    && apt-get -y --no-install-recommends install \
        python3-dev \
        # default-libmysqlclient-dev \
        build-essential \
        ca-certificates
```

Where it's failing is at connect:
```python
        connection = MySQLdb.connect(
            host=os.getenv("HOST"),
            user=os.getenv("USERNAME"),
            passwd=os.getenv("PASSWORD"),
            db=os.getenv("DATABASE"),
            client_flag=CLIENT.SSL,
            ssl_mode="VERIFY_IDENTITY",
            ssl={"ca": os.getenv("SSL_CERT")},
        )
```

I was also looking at the `MySQLdb` source code module and found that the direct cause of the error is the inclusion of the `ssl` dict. It will fail if it detects the installed `mysql` does not support ssl.

**Comment by fooforge** [comment, marked as answer]
Hey @Sammyalhashe,

With Buster, the Debian project seems to have stopped _bundling_ the `mysql` client, but instead now installs the MariaDB mysql client.

```bash
fooforge in 🌐 prometheus in ~ took 6s
❯ apt-cache show default-mysql-client
Package: default-mysql-client
Source: mysql-defaults
Version: 1.0.7
Installed-Size: 10
Maintainer: Debian MySQL Maintainers <pkg-mysql-maint@lists.alioth.debian.org>
Architecture: all
Depends: mariadb-client-10.5
Description-en: MySQL database client binaries (metapackage)
 MySQL is a fast, stable and true multi-user, multi-threaded SQL database
 server. SQL (Structured Query Language) is the most popular database query
 language in the world. The main goals of MySQL are speed, robustness and
 ease of use.
 .
 This package depends on the default implementation of the client binaries and
 the additional tools innotop and mysqlreport.
Description-md5: 0d975e58d81d31c2c1965d43804779c1
Section: database
Priority: optional
Filename: pool/main/m/mysql-defaults/default-mysql-client_1.0.7_all.deb
Size: 3700
MD5sum: 8f5aa5e915506dd8322e04c80659eb3d
SHA256: d772d5e44ba94dd87d7da717c8bf7091fd6ab932d5ab6aee782f133745d2add5
```

That client's options are named slightly differently. I was able to connect to PlanetScale like so:

```bash
fooforge in 🌐 prometheus in ~
❯ mysql -h eu-central.connect.psdb.cloud -u xxxxx -ppscale_pw_xxxxx --ssl-verify-server-cert=TRUE --ssl-ca=/etc/ssl/certs/ca-certificates.crt
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MySQL connection id is 3558481665
Server version: 8.0.23-PlanetScale

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MySQL [(none)]>
```

Now, I cannot say if MySQLdb or the [`mysql-connector-python` connector](https://dev.mysql.com/doc/connector-python/en/) are flexible enough to make use of MariaDB's libraries, but as a workaround you could simply set up MySQL's APT repository and then install the native MySQL client and libraries. Just make sure you don't have the MariaDB client installed anymore, otherwise the installation will fail due to a conflict.

I'm going to log the commands I have run below, but would also recommend to read through [MySQL's official documentation](https://dev.mysql.com/doc/mysql-apt-repo-quick-guide/en/) on their APT repositories.

```bash
fooforge in 🌐 prometheus in ~
❯ curl -OLs https://dev.mysql.com/get/mysql-apt-config_0.8.24-1_all.deb

fooforge in 🌐 prometheus in ~
❯ sudo dpkg -i mysql-apt-config_0.8.24-1_all.deb
...

fooforge in 🌐 prometheus in ~
> sudo apt-get update
...

fooforge in 🌐 prometheus in ~
❯ apt-cache show mysql-client
Package: mysql-client
Source: mysql-community
Version: 8.0.31-1debian11
Architecture: amd64
Maintainer: MySQL Release Engineering <mysql-build@oss.oracle.com>
Installed-Size: 77
Depends: mysql-community-client (= 8.0.31-1debian11)
Homepage: http://www.mysql.com/
Priority: optional
Section: database
Filename: pool/mysql-8.0/m/mysql-community/mysql-client_8.0.31-1debian11_amd64.deb
Size: 68304
SHA256: fd99226e9672c95b2758c640931f96fc2ee957c5578014a926ebcbbf6ab8100b
SHA1: b27e38d31c7917448ef136cb6dab205dd1921fe8
MD5sum: c746fa91320b4479e61596c56effd38a
Description: MySQL Client meta package depending on latest version
 The MySQL(TM) software delivers a very fast, multi-threaded, multi-user,
 and robust SQL (Structured Query Language) database server. MySQL Server
 is intended for mission-critical, heavy-load production systems as well
 as for embedding into mass-deployed software. MySQL is a trademark of
 Oracle. This is a meta package that depends on the latest mysql client
 package available in the repository.
Description-md5: 9562f9a79c0f92ac81175a806e88ebfd

fooforge in 🌐 prometheus in ~
❯ sudo apt-get install mysql-client
...

fooforge in 🌐 prometheus in ~
❯ mysql -h eu-central.connect.psdb.cloud -u xxxxx -ppscale_pw_xxxxx --ssl-mode=VERIFY_IDENTITY --ssl-ca=/etc/ssl/certs/ca-certificates.crt
mysql: [Warning] Using a password on the command line interface can be insecure.
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 129076032
Server version: 8.0.23-PlanetScale

Copyright (c) 2000, 2022, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql>
```

After installing `libmysqlclient-dev` from MySQL's APT repository you should now also be able to connect from your Python application.

**Reply by Sammyalhashe** [reply]
Wow! Thanks for the elaborate response! I will try this when I get back from work and will update how it goes :)

**Reply by Sammyalhashe** [reply]
So as an update I tried it out, ran into an issue where it says it can't find mysql-client (both apt-cache show and apt-get install say this). The issue I believe is that when building a docker container there is a step in dpkg configure where it manually asks you to choose how to configure mysql-apt-config. It gives a prompt like this:

```
Configuring mysql-apt-config
----------------------------

MySQL APT Repo features MySQL Server along with a variety of MySQL components.
You may select the appropriate product to choose the version that you wish to
receive.

Once you are satisfied with the configuration then select last option 'Ok' to
save the configuration, then run 'apt-get update' to load package list.
Advanced users can always change the configurations later, depending on their
own needs.

  1. MySQL Server & Cluster (Currently selected: mysql-8.0)
  2. MySQL Tools & Connectors (Currently selected: Enabled)
  3. MySQL Preview Packages (Currently selected: Disabled)
  4. Ok
Which MySQL product do you wish to configure? 
```
The problem is building a docker container I am by default in a non-interactive mode. Do you know if there is a way around this and I could choose a default configuration?

**Reply by fooforge** [reply]
@Sammyalhashe Yes. We can use `debconf` to make our selections before we install the package. Try this:

```bash
fooforge in 🌐 prometheus in ~
❯ echo mysql-apt-config mysql-apt-config/select-product select Ok | sudo debconf-set-selections

fooforge in 🌐 prometheus in ~
❯ sudo DEBIAN_FRONTEND=noninteractive dpkg -i mysql-apt-config_0.8.24-1_all.deb
(Reading database ... 55318 files and directories currently installed.)
Preparing to unpack mysql-apt-config_0.8.24-1_all.deb ...
Unpacking mysql-apt-config (0.8.24-1) over (0.8.24-1) ...
Setting up mysql-apt-config (0.8.24-1) ...
Warning: apt-key should not be used in scripts (called from postinst maintainerscript of the package mysql-apt-config)
Warning: apt-key is deprecated. Manage keyring files in trusted.gpg.d instead (see apt-key(8)).
OK

fooforge in 🌐 prometheus in ~
❯
```

**Reply by Sammyalhashe** [reply]
Thanks @fooforge! Well, so I was able to install mysql-client! But now during the setup wheel for mysqlclient in python, it fails saying mysql_config is not found.

So far I have installed:
mysql-client
python3-dev
build-essential

To be honest, not sure of what else I could be missing here.

**Reply by Sammyalhashe** [reply]
The alternative for me I guess is to use `mysql-connector-python` instead. That seems to install fine. We shall see if it can connect to planetscale though (considering the original problem this discussion was opened for).

**Reply by fooforge** [reply]
@Sammyalhashe You'll want to add `libmysqlclient-dev` to that list. :-)

```bash
fooforge in 🌐 prometheus in ~
❯ dpkg-query -L libmysqlclient-dev
/.
/usr
/usr/bin
/usr/bin/mysql_config
/usr/include
/usr/include/mysql
...
...
```

**Reply by Sammyalhashe** [reply]
> @Sammyalhashe You'll want `libmysqlclient-dev`. :-)
> 
> 
> 
> ```bash
> 
> fooforge in 🌐 prometheus in ~
> 
> ❯ dpkg-query -L libmysqlclient-dev
> 
> /.
> 
> /usr
> 
> /usr/bin
> 
> /usr/bin/mysql_config
> 
> /usr/include
> 
> /usr/include/mysql
> 
> ...
> 
> ...
> 
> ```
> 
> 

Oh nice :) was working on this before work, will try when I get back. Thanks for all your help!

**Reply by Sammyalhashe** [reply]
So I got it working. Thanks for your patience and help :)

**Reply by fooforge** [reply]
@Sammyalhashe Great. You're welcome!

**Comment by Sammyalhashe** [comment]
Quick follow up question, if I'm getting an operational error that it's unable to connect to the branch xxxxx. does that mean branch xxxxx does not exist anymore in the database we have in planetscale?

**Reply by Sammyalhashe** [reply]
And how can I specify which branch to connect to using python?
